File: examples/python/airtable/airtable.py
# ========= Copyright 2026 @ Strukto.AI All Rights Reserved. =========
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
#     http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.
# ========= Copyright 2026 @ Strukto.AI All Rights Reserved. =========

import asyncio
import os

from dotenv import load_dotenv

from mirage import MountMode, Workspace
from mirage.vfs.airtable import AirtableConfig, AirtableVFS

load_dotenv(".env.development")

config = AirtableConfig(token=os.environ["AIRTABLE_TOKEN"])
vfs = AirtableVFS(config=config)


async def run(ws: Workspace, cmd: str) -> str:
    result = await ws.shell(cmd)
    print(f"$ {cmd}")
    if result.exit_code != 0:
        print(f"  exit={result.exit_code}  "
              f"{(await result.stderr_str()).strip()[:200]}")
    out = (await result.stdout_str()).rstrip()
    for line in out.splitlines()[:10]:
        print(f"  {line[:200]}")
    return out


async def main() -> None:
    ws = Workspace({"/airtable": vfs}, mode=MountMode.READ)
    try:
        bases = (await run(ws, "ls /airtable/bases")).splitlines()
        if not bases:
            print("No bases visible to this token")
            return
        base = f"/airtable/bases/{bases[0]}"
        await run(ws, f"cat {base}/base.json")
        tables = [
            name for name in (await run(ws, f"ls {base}")).splitlines()
            if name != "base.json"
        ]
        if not tables:
            return
        table = f"{base}/{tables[0]}"
        await run(
            ws, f"jq -r '.fields[] | .field_name + \": \" + .type' "
            f"{table}/table.json")
        # head pushes its count into maxRecords: one request, not the table
        await run(ws, f"head -n 3 {table}/records.jsonl | jq -c .fields")
        await run(ws, f"ls {table}/views")
        await run(
            ws, "cat /airtable/bases/Missing__appMissing00000001/"
            "base.json")
    finally:
        await ws.close()


if __name__ == "__main__":
    asyncio.run(main())

File: examples/typescript/airtable/airtable.ts
// ========= Copyright 2026 @ Strukto.AI All Rights Reserved. =========
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//     http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.
// ========= Copyright 2026 @ Strukto.AI All Rights Reserved. =========

import { resolve } from 'node:path'
import { fileURLToPath } from 'node:url'
import dotenv from 'dotenv'
import { AirtableVFS, MountMode, Workspace, type AirtableConfig } from '@struktoai/mirage-node'

const __HERE = fileURLToPath(new URL('.', import.meta.url))
dotenv.config({ path: resolve(__HERE, '../../../.env.development') })

function buildConfig(): AirtableConfig {
  const token = process.env.AIRTABLE_TOKEN
  if (token === undefined || token === '') {
    throw new Error('AIRTABLE_TOKEN env var is required')
  }
  return { token }
}

async function run(ws: Workspace, cmd: string): Promise<string> {
  console.log(`$ ${cmd}`)
  const r = await ws.shell(cmd)
  if (r.exitCode !== 0) {
    console.log(`  exit=${String(r.exitCode)}  ${r.stderrText.trim().slice(0, 200)}`)
  }
  const out = r.stdoutText.replace(/\s+$/, '')
  if (out !== '') {
    for (const line of out.split('\n').slice(0, 10)) console.log(`  ${line.slice(0, 200)}`)
  }
  return out
}

async function main(): Promise<void> {
  const ws = new Workspace({ '/airtable': new AirtableVFS(buildConfig()) }, { mode: MountMode.READ })
  try {
    const bases = (await run(ws, 'ls /airtable/bases')).split('\n').filter((b) => b !== '')
    const first = bases[0]
    if (first === undefined) {
      console.log('No bases visible to this token')
      return
    }
    const base = `/airtable/bases/${first}`
    await run(ws, `cat ${base}/base.json`)
    const tables = (await run(ws, `ls ${base}`))
      .split('\n')
      .filter((name) => name !== '' && name !== 'base.json')
    const firstTable = tables[0]
    if (firstTable === undefined) return
    const table = `${base}/${firstTable}`
    await run(ws, `jq -r '.fields[] | .field_name + ": " + .type' ${table}/table.json`)
    // head pushes its count into maxRecords: one request, not the table
    await run(ws, `head -n 3 ${table}/records.jsonl | jq -c .fields`)
    await run(ws, `ls ${table}/views`)
    await run(ws, 'cat /airtable/bases/Missing__appMissing00000001/base.json')
  } finally {
    await ws.close()
  }
}

await main()
